packages/headless/src/controllers/recent-queries-list/headless-recent-queries-list.ts

```diff
@@ -0,0 +1,148 @@
+import {
+  clearRecentQueries,
+  registerRecentQueries,
+} from '../../features/recent-queries/recent-queries-actions';
+import {
+  validateInitialState,
+  validateOptions,
+} from '../../utils/validate-payload';
+import {executeSearch} from '../../features/search/search-actions';
+import {updateQuery} from '../../features/query/query-actions';
+import {
+  logClearRecentQueries,
+  logRecentQueryClick,
+} from '../../features/recent-queries/recent-queries-analytics-actions';
+
+export interface RecentQueriesListProps {
+  /**
+   * The initial state that should be applied to the `RecentQueriesList` controller.
+   */
+  initialState?: RecentQueriesListInitialState;
+  /**
+   * The configuration options that should be applied to the `RecentQueriesList` controller.
+   */
+  options?: RecentQueriesListOptions;
+}
+
+export interface RecentQueriesListInitialState {
+  /**
+   * The list of recent queries
+   * @defaultValue `[]`
+   */
+  queries: string[];
+}
+
+const initialStateSchema = new Schema<RecentQueriesListInitialState>({
+  queries: new ArrayValue({required: false}),
+});
+
+export interface RecentQueriesListOptions {
+  /**
+   * The maximum number of queries to retain in the list.
+   * @defaultValue `10`
+   */
+  maxQueries: number;
+}
+
+const optionsSchema = new Schema<RecentQueriesListOptions>({
+  maxQueries: new NumberValue({required: false}),
+});
+
+/**
+ * The `RecentQueriesList` controller is in charge of managing the user's recent queries.
+ */
+export interface RecentQueriesList extends Controller {
+  /**
+   * The state of the RecentQueriesList controller.
+   * */
+  state: RecentQueriesState;
+  /**
+   * Clear the recent queries list.
+   */
+  clear(): void;
+  /**
+   * Execute the given recent query.
+   * @param query - The recent query to execute.
+   */
+  executeRecentQuery(query: string): void;
+}
+
+export function validateRecentQueriesProps(
+  engine: SearchEngine,
+  props?: RecentQueriesListProps
+) {
+  validateOptions(
+    engine,
+    optionsSchema,
+    props?.options,
+    'buildRecentQueriesList'
+  );
+  validateInitialState(
+    engine,
+    initialStateSchema,
+    props?.initialState,
+    'buildRecentQueriesList'
+  );
+}
+
+/**
+ * Creates a `RecentQueriesList` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configuration `RecentQueriesList` properties.
+ * @returns A `RecentQueriesList` controller instance.
+ * */
+export function buildRecentQueriesList(
+  engine: SearchEngine,
+  props?: RecentQueriesListProps
+): RecentQueriesList {
+  if (!loadRecentQueriesListReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+  const getState = () => engine.state;
+
+  validateRecentQueriesProps(engine, props);
+  dispatch(
+    registerRecentQueries({
+      queries: props?.initialState?.queries ?? [],
```

Comment:
Instead of this pattern, I would propose having a `defaultRecentQueriesProps` that gets merged with the provided one, with a spread operator.

That way you could remove all the optional, and there would be a single place where we see "here is the default if nothing is provided".